Changed clip_grad_norm_ total_norm calculation

Enealor opened the pull request:
Redefines the computation of the total_norm to increase performance as shown in #31474.

Enealor committed:
Changed total_norm calculation

total_norm for norm_type != inf has been changed to increase speed.

vadimkantorov commented:
The change doesn't use `.data` whereas the original code does. Probably needs `with torch.no_grad()` somewhere or `@torch.no_grad()` (cc @albanD)?

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 33a314c9:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=33a314c9976be4cd99d68939b760f26e716c271b).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4195157) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=82956045) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=82956045))
<details>
<summary>
<code>Jan 10 00:52:00 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 10 00:52:00 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 10 00:52:00 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 10 00:52:00 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 10 00:52:00 1 out of 1 hunk FAILED 
Jan 10 00:52:00 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 10 00:52:00 Traceback (most recent call last): 
Jan 10 00:52:00   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 10 00:52:00     patch_repo(args) 
Jan 10 00:52:00   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 10 00:52:00     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 10 00:52:00 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 10 00:52:00 =================== sccache compilation log =================== 
Jan 10 00:52:00 + cleanup 
Jan 10 00:52:00 + retcode=1 
Jan 10 00:52:00 + set +x 
Jan 10 00:52:00 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 10 00:52:00 Compile requests                 1 
Jan 10 00:52:00 Compile requests executed        0 
Jan 10 00:52:00 Cache hits                       0 
Jan 10 00:52:00 Cache misses                     0 
Jan 10 00:52:00 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Enealor committed:
Updated clip_grad_norm_ computation

Changed p.grad to p.grad.data. Changed computation from sum(x**p)**(1/p) to norm(x, p).

Enealor commented:
I changed the expression to use `p.grad.data` to make it more similar to the previous and surrounding code. I traded the powers and sums used previously for a `torch.norm` to make it easier to read. It appears to have a positive impact on performance as well.

albanD commented:
As mentionned by @vadimkantorov  you want to either:
- replace the `.data` with a `.detach()`.
- wrap your computations in a `with torch.no_grad():` block and remove the `.data`.

I am a bit worried about the memory consumption here though. This will re-allocate a buffer as big as all the parameters. That might not be great for performance either.
Does allocating a buffer of size `len(parameters)` (on the gpu), computing all the intermediary losses into it and then summing it be enough to reduce the overhead of the multiple GPU-CPU copies while keeping the memory use low? cc @ngimel

ngimel commented:
Memory consumption is ok, you are stacking already reduced parameters, so extra memory = number of parameters. The original performance bottleneck was not because of cpu-gpu copies (there were none), but because of pow and add operations after every reduce operation, but it does not matter, present approach fixes it anyway. 
@enealor, it will be great if you can provide benchmarks for before and after. I had a crude script initially, comparing original implementation and the one proposed in #31474 for an imitation of a network consisting of a few linear layers. It's not a clean comparison (in particular, it does not use .grad and directly computes norm for a list of tensors) so if you have better benchmarking script, please post it on gist, and link it and benchmarking results here, in the PR. 
```
import torch
import time


nlayers = 10
params = []
size = 1024
for _ in range(nlayers):
    params.append(torch.randn(size, device="cuda", requires_grad=True))
    params.append(torch.randn((size, size), device="cuda", requires_grad=True))


def bench(fn, nrep, a):
    #dry tun
    c=fn(a)
    torch.cuda.synchronize()
    start = time.time()
    for _ in range(nrep):
        c=fn(a)
    torch.cuda.synchronize()
    end = time.time()
    return((end-start)/nrep)



def norm_orig(parameters):
    total_norm = 0.0
    norm_type = 2
    for p in parameters:
        param_norm = p.data.norm(2)
        total_norm += param_norm ** norm_type
    return total_norm

def norm_1(parameters):
    norm_type = 2
    total_norm = torch.sum(torch.stack([torch.norm(p.data,norm_type) for p in parameters]) ** norm_type)
    return total_norm

nrep = 100
t0 = bench(norm_orig, nrep, params)
t1 = bench(norm_1, nrep, params)

print(t0, t1)
```

SsnL commented on torch/nn/utils/clip_grad.py:
```diff
@@ -27,11 +27,7 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2):
-            param_norm = p.grad.data.norm(norm_type)
-            total_norm += param_norm.item() ** norm_type
-        total_norm = total_norm ** (1. / norm_type)
+        total_norm = torch.norm(torch.stack([torch.norm(p.grad.data, norm_type) for p in parameters]), norm_type)
```
wait, norm of norms of segments is not generally equivalent with norm, right?

ngimel replied:
I stopped there for a moment too, but turns out that it is.

albanD replied:
I though the same, but for 2 norm it works.
For two segments a and b: `(a.pow(2).sum().sqrt().pow(2) + b.pow(2).sum().sqrt().pow(2)).sqrt() = (a.pow(2).sum() + b.pow(2).sum()).sqrt()`.
Wouldn't work for squared norm though.

SsnL replied:
Hmm don't you need a `pow` here for this to work though?

albanD replied:
The second norm is doing it.

SsnL replied:
ohhh right. thanks for the explanation!

Enealor replied:
I checked the math a few times to be sure that this is correct. In fact, this motivated one of my first suggestions for improving the performance, as we potentially spend time taking a root and then raising it to the same power. (That suggestion didn't work due to memory issues).

If it helps, [this script](https://gist.github.com/Enealor/b52e0cf7a67803ccff006f8e6dd6ada9) prints the result of computing the norm in each way. You can experimentally verify it works correctly for other powers quickly this way.

Enealor commented:
I used both a variant of the script you posted and a custom script. I changed yours to include the new version I suggested, and also completed the computation of the total norm to get a fairer comparison.

An example output of the [profile script](https://gist.github.com/Enealor/ed17d56ad4851f10182c32e84c5d947f) I wrote:
```
(torchnl) PS D:\Code\gradtest> python profile_total_norm.py --calls 500 --device cpu
Parameters: 9550849
Time spent in total_norm_1: 22.2670s total, 0.04453s per call
Time spent in total_norm_2: 21.5541s total, 0.04311s per call
Time spent in total_norm_3: 20.8662s total, 0.04173s per call
(torchnl) PS D:\Code\gradtest> python profile_total_norm.py --calls 500 --device cuda
Parameters: 9550849
Time spent in total_norm_1: 2.7180s total, 0.00544s per call
Time spent in total_norm_2: 1.2507s total, 0.00250s per call
Time spent in total_norm_3: 1.3107s total, 0.00262s per call
```
(`total_norm_1` is the original version. `total_norm_2` is the first revision. `total_norm_3` is my revision.)

An example output of the [modified profile script](https://gist.github.com/Enealor/ed17d56ad4851f10182c32e84c5d947f):
```
(torchnl) PS D:\Code\gradtest> python .\norm_bench.py
3239.947021484375
3239.947265625
3239.947265625
0.004649996757507324 0.002029845714569092 0.001959977149963379
1.0 0.43652626451662524 0.42150075627451483
```
The first three lines are the computed norms. The fourth line is the time per call, in the same order as the previous script. The last line is the ratio of the time in each call with the time spent in the original version.

Enealor committed:
Replaced .data with .detach()

Updated references to the .data attribute with .detach() calls in the clip_grad_norm_ function.

Enealor commented:
In response to the comments on using `.data`, I updated the references in the function to `.detach()`.

ezyang commented:
Flagging @dlibenzi @ailzhang, this broke the XLA build. Do you want me to revert this or is the fix easy?

dlibenzi commented:
Fixing it...

ngimel commented:
This also broke internal tests, so I'm going to add .item() to the computed norm - hopefully this will fix XLA too?

Enealor commented:
It doesn't look like the XLA issue is due to the code itself, but due to an XLA patch that is no longer up to date with this change. In particular, the Circle CI points out that `RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff`, and assuming that this refers to [this patch](https://github.com/pytorch/xla/blob/b6fad06d2c29d56f87d539d80af4ac60993f3a25/torch_patches/X10-clip_grad.diff), then I think it didn't work because it was trying to change code that no longer exists.

On one hand, dlibenzi already has the patch more up to date so that it reverts the new code to the previous form. However, did people get a chance to check if the code was the issue, and not the failure to apply the patch?

ngimel commented:
#32159 to fix return type

dlibenzi commented:
Note that XLA patch has already been fixed and pushed yesterday.
So if this changes again, we will likely have to refresh the patch once again.

stephenroller commented:
I think this broke clipping gradient norms for models with parameters on more than one GPU.
